src/vuln_analysis/utils/java_segmenters_with_methods.py

```diff
@@ -0,0 +1,1749 @@
+    return i
+
+# --------------- delimiter & ignored-span index (one-pass pre-scan) ----------
+
+_PAR_O2C: Dict[int, int] = {}
+_PAR_C2O: Dict[int, int] = {}
+_BR_O2C: Dict[int, int] = {}
+_BR_C2O: Dict[int, int] = {}
+
+# All spans to ignore for scanning (comments, strings, text blocks)
+_IGN_SPANS: List[Tuple[int, int]] = []   # [(start, end)], end exclusive
+_IGN_STARTS: List[int] = []
+_IGN_ENDS: List[int] = []
+
+_DELIM_SRC_ID: Optional[int] = None
+
+def _ignored_span_at(pos: int) -> Optional[Tuple[int, int]]:
+    """Return (start,end) ignored span containing pos, or None."""
+    if not _IGN_SPANS:
+        return None
+    idx = bisect.bisect_right(_IGN_STARTS, pos) - 1
+    if idx >= 0 and _IGN_STARTS[idx] <= pos < _IGN_ENDS[idx]:
+        return _IGN_SPANS[idx]
+    return None
+
+def _build_delim_index(src: str) -> Tuple[
+    Dict[int, int], Dict[int, int], Dict[int, int], Dict[int, int], List[Tuple[int, int]]
+]:
+    par_o2c: Dict[int, int] = {}
+    par_c2o: Dict[int, int] = {}
+    br_o2c: Dict[int, int] = {}
+    br_c2o: Dict[int, int] = {}
+
+    spans: List[Tuple[int, int]] = []
+    stack_par: List[int] = []
+    stack_br: List[int] = []
+
+    i, n = 0, len(src)
+    while i < n:
+        ch = src[i]
+
+        # whitespace
+        if ch in (' ', '\t', '\r', '\n', '\f'):
+            i += 1
+            continue
+
+        # line comment
+        if i + 1 < n and ch == '/' and src[i + 1] == '/':
+            start = i
+            i += 2
+            nl = src.find('\n', i)
+            i = n if nl == -1 else nl + 1
+            spans.append((start, i))
+            continue
+
+        # block comment (incl. Javadoc)
+        if i + 1 < n and ch == '/' and src[i + 1] == '*':
+            start = i
+            j = src.find('*/', i + 2)
+            i = n if j == -1 else j + 2
+            spans.append((start, i))
+            continue
+
+        # strings / chars / text blocks
+        if src.startswith('"""', i) or ch in ('"', "'"):
+            start = i
+            i = _skip_java_string_like(src, i)
+            spans.append((start, i))
+            continue
+
+        # delimiters
+        if ch == '(':
+            stack_par.append(i)
+        elif ch == ')' and stack_par:
+            o = stack_par.pop()
+            par_o2c[o] = i
+            par_c2o[i] = o
+        elif ch == '{':
+            stack_br.append(i)
+        elif ch == '}' and stack_br:
+            o = stack_br.pop()
+            br_o2c[o] = i
+            br_c2o[i] = o
+
+        i += 1
+
+    spans.sort(key=lambda t: t[0])
+    return par_o2c, par_c2o, br_o2c, br_c2o, spans
+
+def _set_delim_index(src: str) -> None:
+    global _PAR_O2C, _PAR_C2O, _BR_O2C, _BR_C2O, _DELIM_SRC_ID, _IGN_SPANS, _IGN_STARTS, _IGN_ENDS
+    _PAR_O2C, _PAR_C2O, _BR_O2C, _BR_C2O, spans = _build_delim_index(src)
+    _IGN_SPANS = spans
+    _IGN_STARTS = [a for (a, _) in spans]
+    _IGN_ENDS = [b for (_, b) in spans]
+    _DELIM_SRC_ID = id(src)
```

Comment:
@tmihalac Please add doumentation to function, as it's really hard to read -> purpose, input, and since there is no output, how the input is being changed. It's also will be more clear if you will return an new string after all the transformation and manipulations finished.

Reply:
OK